master/buildbot/reporters/slack.py

```diff
@@ -0,0 +1,177 @@
+# This file is part of Buildbot.  Buildbot is free software: you can
+# redistribute it and/or modify it under the terms of the GNU General Public
+# License as published by the Free Software Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+"""
+Push build statuses to Slack
+"""
+from twisted.internet import defer
+
+from buildbot import config
+from buildbot.process.results import SUCCESS
+from buildbot.reporters.http import HttpStatusPushBase
+from buildbot.util.httpclientservice import HTTPClientService
+from buildbot.util.logger import Logger
+
+
+_LOG = Logger()
+
+
+class _Default(object):
+    # pylint: disable=too-few-public-methods
+    """
+    this is a sentinel value for the default_channels parameter below
+    """
+
+
+class SlackReporter(HttpStatusPushBase):
+    # pylint: disable=invalid-name, too-many-ancestors
+    """
+    minimal slack integration
+    """
+    def checkConfig(self,
+                    username,   # pylint: disable=unused-argument
+                    url,        # pylint: disable=unused-argument
+                    mapping,
+                    default_channels=_Default,
+                    *args, **kwargs):
+        """
+        check if the provided configuration is OK
+        """
+        if not isinstance(mapping, dict):
+            config.error('mapping parameter must be a dictionary')
+        if not (default_channels is _Default or
+                isinstance(default_channels, (list, tuple))):
+            config.error('default_channels parameter must be a list or a tuple')
+
+        super(SlackReporter).checkConfig(*args, **kwargs)
+
+    @defer.inlineCallbacks
+    def reconfigService(self, username, url, mapping, default_channels,
+                        *args, **kwargs):
+        """
+        reconfigure the service based on the new parameters
+        """
+        yield super(SlackReporter).reconfigService(self, *args, **kwargs)
+
+        # pylint: disable=attribute-defined-outside-init
+
+        self._username = username
+
+        # TODO: this goes to the documentation
+        # mapping is expected to have the following structure
+        # mapping = {
+        #     'builder1': {
+        #         'kinds': ('start',),
+        #         'channels': ('#channel1', '#channel2')
+        #     },
+        #     'builder2': {
+        #         'kinds': ('finish',),
+        #     },
+        #     'builder3': {
+        #         'kinds': ('start', 'finish')
+        #     },
+        # }
```

Comment:
Another though: for e.g dan who as 4000 builders, that mapping would be painful to write.

I think the http base reporter already has some kind of builder filtering. 
The HipChat reporter has another filtering semantics. while it is also a fairly straightforward chat system.
Buildbot will clearly benefit to have similar configurability for all chat services.

Shall we make a HttpChatReporterBase class (or a ChatReporterMixin, which would choose the user/rooms where to post) ?